Automattic/Gravatar-SDK-iOS
# Refactor image cache

Closes #157 

### Description

This PR is an experiment implementing the cache system described [here.](https://developer.apple.com/tutorials/app-dev-training/caching-network-data).

There are two main changes:

- `ImageCache` is now an actor.
This serialises access to the internal NSCache instance.

- ImageCache will also cache the tasks of getting an image from remote.
This will avoid executing the same network call many times while waiting for a response.

Currently on the demo app `Image Download` screen, if you tap on `Fetch Avatar` many times, all those taps will turn into a network call, until any of the created network calls return.
With this change, there will be just one network call, and all other taps will await for this network call to return.

### Testing Steps

- Setup Proxyman or similar.
- Run the demo app and access `Image download - Networking` screen.
- Insert a valid Gravatar email account.
- Tap the fetch image button many times fast.
- Check on Proxyman that there's only one network call excecuted.
